base/sparse/sparsematrix.jl

```diff
@@ -1884,48 +1884,61 @@ function setindex!{Tv,Ti,T<:Real}(A::SparseMatrixCSC{Tv,Ti}, x, I::AbstractVecto
 end
 
 
-
 # Sparse concatenation
 
 function vcat(X::SparseMatrixCSC...)
+    Tv = promote_type(map(x->eltype(x.nzval), X)...)
+    Ti = promote_type(map(x->eltype(x.rowval), X)...)
+
+    vcat(map(x->convert(SparseMatrixCSC{Tv,Ti}, x), X)...)
```

Comment:
I am not really familiar with the sparse code, but I agree that @timholy is probably right. It should not be necessary to convert a complete array. By allocating the output array with the promoted type and then assigning into it, the convert should be taken care during the assignment at the level of single elements.